tiered-storage/jcloud/src/main/java/org/apache/bookkeeper/mledger/offload/jcloud/impl/BufferedOffloadStream.java

```diff
@@ -0,0 +1,148 @@
+    static final int ENTRY_HEADER_SIZE = 4 /* entry size */ + 8 /* entry id */;
+    private final long blockSize;
+    private final ConcurrentLinkedQueue<Entry> entryBuffer;
+    private final InputStream blockHead;
+    int offset = 0;
+    static int NOT_INITIALIZED = -1;
+    int validDataOffset = NOT_INITIALIZED;
+    CompositeByteBuf currentEntry;
+
+    public long getLedgerId() {
+        return ledgerId;
+    }
+
+    public long getBeginEntryId() {
+        return beginEntryId;
+    }
+
+    public long getBlockSize() {
+        return blockSize;
+    }
+
+
+    public BufferedOffloadStream(int blockSize,
+                                 ConcurrentLinkedQueue<Entry> entryBuffer,
+                                 SegmentInfoImpl segmentInfo,
+                                 long ledgerId,
+                                 long beginEntryId,
+                                 AtomicLong bufferLength) {
+        this.ledgerId = ledgerId;
+        this.beginEntryId = beginEntryId;
+        this.blockSize = blockSize;
+        this.segmentInfo = segmentInfo;
+        this.entryBuffer = entryBuffer;
+        this.bufferLength = bufferLength;
+        this.blockHead = StreamingDataBlockHeaderImpl.of(blockSize, ledgerId, beginEntryId)
+                .toStream();
+    }
+
+
+    @Override
+    public int read() throws IOException {
+        if (blockHead.available() > 0) {
+            offset++;
+            return blockHead.read();
+        }
+        //if current exists, use current first
+        if (currentEntry != null) {
+            if (currentEntry.readableBytes() > 0) {
+                offset += 1;
+                return currentEntry.readUnsignedByte();
+            } else {
+                currentEntry.release();
+                currentEntry = null;
+            }
+        }
+
+        if (blockSize <= offset) {
+            return -1;
+        } else if (validDataOffset != NOT_INITIALIZED) {
+            return BLOCK_END_PADDING[(offset++ - validDataOffset) % BLOCK_END_PADDING.length];
+        }
+
+        Entry headEntry;
+
+        while ((headEntry = entryBuffer.peek()) == null) {
+            if (segmentInfo.isClosed()) {
+                if (validDataOffset == NOT_INITIALIZED) {
+                    validDataOffset = offset;
+                }
+                return read();
+            } else {
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException e) {
+                    log.error("sleep failed", e);
+                }
+            }
+        }
+
+        //create new block when a ledger end
+        if (headEntry.getLedgerId() != this.ledgerId) {
+            if (validDataOffset == NOT_INITIALIZED) {
+                validDataOffset = offset;
+            }
+            return read();
+        }
+
+        if (blockSize >= offset
+                + ENTRY_HEADER_SIZE
+                + headEntry.getLength()) {
+            entryBuffer.poll();
```

Comment:
We should to recycle the entries after read complete?

Reply:
headEntry.release added